aiohttp/cookiejar.py

```diff
@@ -167,11 +159,40 @@ def __len__(self) -> int:
         return sum(len(cookie.values()) for cookie in self._cookies.values())
 
     def _do_expiration(self) -> None:
-        self.clear(lambda x: False)
+        expire_heap_len = len(self._expire_heap)
+        if not expire_heap_len:
+            return
+
+        if expire_heap_len > 100 and expire_heap_len > len(self._expirations) * 2:
+            self._expire_heap = [
+                entry
+                for entry in self._expire_heap
+                if self._expirations.get(entry[1]) != entry[0]
+            ]
+            heapq.heapify(self._expire_heap)
+
+        now = time.time()
+        to_del: List[Tuple[str, str, str]] = []
+        while self._expire_heap:
+            when, cookie_key = self._expire_heap[0]
+            if when > now:
+                break
+            heapq.heappop(self._expire_heap)
+            if self._expirations.get(cookie_key) == when:
```

Comment:
Needs a comment explaining how it could still be in the heap but not reached and the expire time changes so we don’t want to delete it until we get far enough down in the heap that we hit the new expire time